### PR TITLE
S251 ingest

### PR DESCRIPTION
**Pull request by patrick-troy**

### CODE REVIEW

File: liiatools/datasets/s251/data_generator/sample_data.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+
+def generate_sample_s251_file():
+    """
+    Generate a sample s251 file
+
+    :return: stream of generators containing information required to create a csv file
+    """
+    yield StartTable(expected_columns=s251_columns)
+    for _ in range(10000):
```
Comment (cyramic) [inline liiatools/datasets/s251/data_generator/sample_data.py:49]: Should this be an option sent to the function instead of hard-coded?

Comment (cyramic) [reply liiatools/datasets/s251/data_generator/sample_data.py:49]: (the 10,000 value)

Comment (patrick-troy) [reply liiatools/datasets/s251/data_generator/sample_data.py:49]: Yeah it definitely should. Although the generator stuff in this PR is a bit outdated now and there's a new branch which is more in keeping with how we should be generating data. I'm happy just to remove all the data generator stuff from this PR

**Review by MichaelHanksSF — CHANGES_REQUESTED** [review]
Hello, few things not working at present.
**Logs**
Totally fine files produce empty error logs - do we want this?
**Mandatory fields**
Blank entries for mandatory fields are not generating error logs notifying this
**Missing columns**
When a column is missing, a haystack is produced with attribute error (I'll send this to you)
**File too old**
When a file has end dates that are too old, the error message produced says that Placement end date column is missing, rather than that the entries in Placement end date are too old
**Type error in placement end date**
If you give an invalid entry in a single row of placement end date, the file isn't processed and the error message says that placement end date is missing
**Type errors on currency**
The currency fields have no validation, so you can write any old rubbish in and it comes out the other end
**Numerical fields allow negatives**
This is probably true of 903 and all the others, but we should bottom the numeric fields with 0

**Comment by patrick-troy** [comment]
> Hello, few things not working at present. **Logs** Totally fine files produce empty error logs - do we want this? **Mandatory fields** Blank entries for mandatory fields are not generating error logs notifying this **Missing columns** When a column is missing, a haystack is produced with attribute error (I'll send this to you) **File too old** When a file has end dates that are too old, the error message produced says that Placement end date column is missing, rather than that the entries in Placement end date are too old **Type error in placement end date** If you give an invalid entry in a single row of placement end date, the file isn't processed and the error message says that placement end date is missing **Type errors on currency** The currency fields have no validation, so you can write any old rubbish in and it comes out the other end **Numerical fields allow negatives** This is probably true of 903 and all the others, but we should bottom the numeric fields with 0

**Review by MichaelHanksSF — CHANGES_REQUESTED** [review]
Hello, thanks for making these changes. There are a couple more issues to fix please:

**Currency format**
When a file is submitted with the cost fields as currency, the haystack is still produced due to encoding. We need this to fail more gracefully.
**Mandatory fields**
Apologies, this is probably me be sloppy suggesting which fields need to be mandatory. Ofsted URN is not mandatory (it can't be); placement end date is still showing as mandatory - this shouldn't be. I think it may be a general problem with your field type *date-blank in the schema as the same issue is true of other non-mandatory date fields

**Nice to have**
**When the file is too old**
The error message works here - is it possible for the error message to reference what the retention policy is for the dataset i.e. 6 years or 12 years or whatever
**Missing columns**
This error works just fine. I notice that a file that is missing columns and has blank values for mandatory fields will also give the blank count error message. However, should the code get to the point of evaluating this if the file didn't have the right columns? Shouldn't it stop earlier than this?

**Review by MichaelHanksSF — CHANGES_REQUESTED** [review]
I think there's still a little work to do here.

With the change that now means that files from 2017-2022 will be returned as they are not expected, this is fine in that it works. And if you give a file from 2016 or older the data retention error will kick in instead. However, going forward, will this not need to be manually revised in order to continue providing the right error? Is there a way for it to be able to get the right error through non-hard-coded logic, such that as we move forward, there'll be no need for edits to the code?

la-agg and pan-agg functionality tested and working as expected

**Review by MichaelHanksSF — CHANGES_REQUESTED** [review]
Unfortunately, for files with data in the 2021-22 FY, I am getting the haystack error again:
"FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\Michael.Hanks\\OneDrive - Social Finance Ltd\\Documents\\Practical skills\\Python Scripts\\liia-tools\\liiatools\\spec\\s251\\S251_schema_2022.yml'"

It looks to me like this is only happening for 2021-22 FY files. Here's the file I tested.
[S251_2021.csv](https://github.com/SocialFinanceDigitalLabs/liia-tools/files/12440875/S251_2021.csv)

**Review by MichaelHanksSF — APPROVED** [review]

**Pull request merged**